torch/utils/_sympy/functions.py

```diff
@@ -841,6 +841,141 @@ def _find_localzeros(cls, values, **options):
     _eval_is_zero = lambda s: _torf(i.is_zero for i in s.args)  # noqa: E731
 
 
+class CustomAdd(sympy.core.add.Add):
+
+    """
+    Custom addition class inheriting from sympy.core.add.Add used to optimize
+    incremental binary summations of certain properties.
+    """
+
+    # Whether the expression represented by this is an ordered_summation_of_unique_symbols.
+    # This optimize constructing expressions of the form
+    # a + b + d +  e ..
+    # since for those all we need todo is to order args and we know Add can't do more
+    # optimizations than that.
+    ordered_summation_of_unique_symbols = False
+
+    # TODO 1 support trail constant term i.e: a + b + 10.
+    # TODO 2 support constant coefficient that is the same for all. i.e: 10a + 10b + 10d...
+
+    def __new__(cls, *args, **kwargs):
+        """
+        Override __new__ method to customize creation of CustomAdd instance.
+        """
+        from sympy.core.basic import _args_sortkey
+
+        import torch._dynamo.config as config
+
+        # This way we make sure the optimizations are only triggered when we pass optimize_incremental_summations.
+        # sometimes sympy will call this function during its simplifications for example, in that case we bypass to add.
+        bypass_to_add = not kwargs.get("optimize_incremental_summations", False) 
+        if bypass_to_add:
+            return sympy.Add(*args, **kwargs)
+
+        assert len(args) == 2
+        assert len(kwargs) == 1
+
+        lhs = args[0]
+        rhs = args[1]
+
+        def make_optimized(ordered_args_as_list):
+            obj = super(CustomAdd, cls).__new__(
+                cls, *ordered_args_as_list, evaluate=False
+            )
+            assert isinstance(obj, CustomAdd)
+            obj.ordered_summation_of_unique_symbols = True
+
+            # if True : # config.run_extra_validations: # need to rebase on another PR to see this lol
```

Comment:
will rebase and uncomment this once the https://github.com/pytorch/pytorch/pull/139741 lands